From 70bc775aaa03b30b6b061526412c47784fb05314 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Sat, 3 Aug 2013 17:07:56 -0500
Subject: [PATCH] Use a ConcurrentHashMap for storing class name -> class
 mappings

Plugins that use multiple threads and try to load classes at the same time can deadlock the entire server

diff --git a/src/main/java/org/bukkit/plugin/java/JavaPluginLoader.java b/src/main/java/org/bukkit/plugin/java/JavaPluginLoader.java
index ea30d83..1721d04 100644
--- a/src/main/java/org/bukkit/plugin/java/JavaPluginLoader.java
+++ b/src/main/java/org/bukkit/plugin/java/JavaPluginLoader.java
@@ -61,7 +61,7 @@ public class JavaPluginLoader implements PluginLoader {
     @Deprecated
     protected final Pattern[] fileFilters = fileFilters0;
 
-    private final Map<String, Class<?>> classes0 = new HashMap<String, Class<?>>();
+    private final Map<String, Class<?>> classes0 = new java.util.concurrent.ConcurrentHashMap<String, Class<?>>();
     /**
      * @deprecated Internal field that wasn't intended to be exposed
      */
-- 
1.7.9.6 (Apple Git-31.1)

